### PR TITLE
Update auditd tests

### PR DESCRIPTION
**Pull request by Lopuiz**
|  Related issue|
| --- |
| [279](https://github.com/wazuh/wazuh-qa/issues/279) |

Hello team,

This PR contains the following test:
1. Remove the monitored folder and check if the rule is removed and re-add the folder. The rule must be re-added. 
2. Remove auditd rule five times and check Wazuh ignores folder.
3. Restart auditd and check Wazuh can reconnects it.
4. Check if whodata changes to realtime if auditd is not installed. 

I enclose some of Syscheck's events and Pytest's outputs below.

Best regards,
Eva

***

**Test 1**

Pytest's output:
```
[root@agent-master-1 test_fim]# python3 -m pytest test_audit/test_audit.py::test_remove_and_readd_folder
========================================== test session starts ===========================================
platform linux -- Python 3.7.4, pytest-5.3.1, py-1.8.0, pluggy-0.13.1
rootdir: /vagrant/wazuh-qa/test_wazuh, inifile: pytest.ini
collected 3 items                                                                                        

test_audit/test_audit.py .ss                                                                       [100%]

================================ 1 passed, 2 skipped in 84.56s (0:01:24) =================================
[root@agent-master-1 test_fim]# 
```

And `ossec.log` logs:
```
2019/12/11 14:30:39 ossec-syscheckd[10362] syscheck_audit.c:1376 at filterkey_audit_events(): DEBUG: (6251): Match audit_key: 'key="wazuh_fim"'
2019/12/11 14:30:39 ossec-syscheckd[10362] syscheck_audit.c:628 at audit_parse(): INFO: (6027): Monitored directory '/testdir1' was removed: Audit rule removed.
2019/12/11 14:30:56 ossec-syscheckd[10362] syscheck_audit.c:1090 at audit_reload_rules(): DEBUG: (6275): Reloading Audit rules...
2019/12/11 14:30:56 ossec-syscheckd[10362] syscheck_audit.c:230 at add_audit_rules_syscheck(): DEBUG: (6322): Reloaded audit rule for monitoring directory: '/testdir1'.
2019/12/11 14:30:56 ossec-syscheckd[10362] syscheck_audit.c:1092 at audit_reload_rules(): DEBUG: (6276): Audit rules reloaded. Rules loaded: 3.
2019/12/11 14:30:57 ossec-syscheckd[10362] syscheck_audit.c:1376 at filterkey_audit_events(): DEBUG: (6251): Match audit_key: 'key="wazuh_fim"'
```

**Test 2**

Pytest's output:
```
[root@agent-master-1 test_fim]# python3 -m pytest test_audit/test_audit.py::test_remove_and_readd_folder
========================================== test session starts ===========================================
platform linux -- Python 3.7.4, pytest-5.3.1, py-1.8.0, pluggy-0.13.1
rootdir: /vagrant/wazuh-qa/test_wazuh, inifile: pytest.ini
collected 3 items                                                                                        

test_audit/test_audit.py .ss                                                                       [100%]

================================ 1 passed, 2 skipped in 85.08s (0:01:25) =================================
[root@agent-master-1 test_fim]# 
```

Logs in `ossec.log`:
```
2019/12/12 10:31:17 ossec-syscheckd[13968] syscheck_audit.c:1376 at filterkey_audit_events(): DEBUG: (6251): Match audit_key: 'key="wazuh_fim"'
2019/12/12 10:31:17 ossec-syscheckd[13968] syscheck_audit.c:634 at audit_parse(): WARNING: (6911): Detected Audit rules manipulation: Audit rules removed.
2019/12/12 10:31:17 ossec-syscheckd[13968] syscheck_audit.c:1090 at audit_reload_rules(): DEBUG: (6275): Reloading Audit rules...
2019/12/12 10:31:17 ossec-syscheckd[13968] syscheck_audit.c:230 at add_audit_rules_syscheck(): DEBUG: (6322): Reloaded audit rule for monitoring directory: '/testdir1'.
2019/12/12 10:31:17 ossec-syscheckd[13968] syscheck_audit.c:1092 at audit_reload_rules(): DEBUG: (6276): Audit rules reloaded. Rules loaded: 3.
2019/12/12 10:31:17 ossec-syscheckd[13968] syscheck_audit.c:1376 at filterkey_audit_events(): DEBUG: (6251): Match audit_key: 'key="wazuh_fim"'
2019/12/12 10:31:18 ossec-syscheckd[13968] syscheck_audit.c:1376 at filterkey_audit_events(): DEBUG: (6251): Match audit_key: 'key="wazuh_fim"'
2019/12/12 10:31:18 ossec-syscheckd[13968] syscheck_audit.c:634 at audit_parse(): WARNING: (6911): Detected Audit rules manipulation: Audit rules removed.
2019/12/12 10:31:18 ossec-syscheckd[13968] syscheck_audit.c:1090 at audit_reload_rules(): DEBUG: (6275): Reloading Audit rules...
2019/12/12 10:31:18 ossec-syscheckd[13968] syscheck_audit.c:230 at add_audit_rules_syscheck(): DEBUG: (6322): Reloaded audit rule for monitoring directory: '/testdir1'.
2019/12/12 10:31:18 ossec-syscheckd[13968] syscheck_audit.c:1092 at audit_reload_rules(): DEBUG: (6276): Audit rules reloaded. Rules loaded: 3.
2019/12/12 10:31:19 ossec-syscheckd[13968] syscheck_audit.c:1376 at filterkey_audit_events(): DEBUG: (6251): Match audit_key: 'key="wazuh_fim"'
2019/12/12 10:31:20 ossec-syscheckd[13968] syscheck_audit.c:1376 at filterkey_audit_events(): DEBUG: (6251): Match audit_key: 'key="wazuh_fim"'
2019/12/12 10:31:20 ossec-syscheckd[13968] syscheck_audit.c:634 at audit_parse(): WARNING: (6911): Detected Audit rules manipulation: Audit rules removed.
2019/12/12 10:31:20 ossec-syscheckd[13968] syscheck_audit.c:1090 at audit_reload_rules(): DEBUG: (6275): Reloading Audit rules...
2019/12/12 10:31:20 ossec-syscheckd[13968] syscheck_audit.c:230 at add_audit_rules_syscheck(): DEBUG: (6322): Reloaded audit rule for monitoring directory: '/testdir1'.
2019/12/12 10:31:20 ossec-syscheckd[13968] syscheck_audit.c:1092 at audit_reload_rules(): DEBUG: (6276): Audit rules reloaded. Rules loaded: 3.
2019/12/12 10:31:20 ossec-syscheckd[13968] syscheck_audit.c:1376 at filterkey_audit_events(): DEBUG: (6251): Match audit_key: 'key="wazuh_fim"'
2019/12/12 10:31:21 ossec-syscheckd[13968] syscheck_audit.c:1376 at filterkey_audit_events(): DEBUG: (6251): Match audit_key: 'key="wazuh_fim"'
2019/12/12 10:31:21 ossec-syscheckd[13968] syscheck_audit.c:634 at audit_parse(): WARNING: (6911): Detected Audit rules manipulation: Audit rules removed.
2019/12/12 10:31:21 ossec-syscheckd[13968] syscheck_audit.c:1090 at audit_reload_rules(): DEBUG: (6275): Reloading Audit rules...
2019/12/12 10:31:21 ossec-syscheckd[13968] syscheck_audit.c:230 at add_audit_rules_syscheck(): DEBUG: (6322): Reloaded audit rule for monitoring directory: '/testdir1'.
2019/12/12 10:31:21 ossec-syscheckd[13968] syscheck_audit.c:1092 at audit_reload_rules(): DEBUG: (6276): Audit rules reloaded. Rules loaded: 3.
2019/12/12 10:31:22 ossec-syscheckd[13968] syscheck_audit.c:1376 at filterkey_audit_events(): DEBUG: (6251): Match audit_key: 'key="wazuh_fim"'
2019/12/12 10:31:23 ossec-syscheckd[13968] syscheck_audit.c:1376 at filterkey_audit_events(): DEBUG: (6251): Match audit_key: 'key="wazuh_fim"'
2019/12/12 10:31:23 ossec-syscheckd[13968] syscheck_audit.c:634 at audit_parse(): WARNING: (6911): Detected Audit rules manipulation: Audit rules removed.
2019/12/12 10:31:23 ossec-syscheckd[13968] syscheck_audit.c:1154 at audit_main(): DEBUG: (6277): Audit thread finished.
2019/12/12 10:31:23 ossec-syscheckd[13968] syscheck_audit.c:1357 at clean_rules(): DEBUG: (6250): Deleting Audit rules...
```

**Test 3**

Pytest's output:
```
[root@agent-master-1 test_fim]# python3 -m pytest test_audit/test_audit.py::test_reconnect_to_audit
========================================== test session starts ===========================================
platform linux -- Python 3.7.4, pytest-5.3.1, py-1.8.0, pluggy-0.13.1
rootdir: /vagrant/wazuh-qa/test_wazuh, inifile: pytest.ini
collected 3 items                                                                                        

test_audit/test_audit.py .ss                                                                       [100%]

================================ 1 passed, 2 skipped in 69.26s (0:01:09) =================================
[root@agent-master-1 test_fim]# 
```

Logs in `ossec.log`:
```
2019/12/12 13:01:10 ossec-syscheckd[26029] syscheck_audit.c:1263 at audit_read_events(): WARNING: (6912): Audit: connection closed.
2019/12/12 13:01:11 ossec-syscheckd[26029] syscheck_audit.c:1267 at audit_read_events(): INFO: (6029): Audit: reconnecting... (1)
2019/12/12 13:01:11 ossec-syscheckd[26029] syscheck_audit.c:199 at init_auditd_socket(): ERROR: (6636): Cannot connect to socket '/var/ossec/queue/ossec/audit'.
2019/12/12 13:01:11 ossec-syscheckd[26029] syscheck_audit.c:1270 at audit_read_events(): INFO: (6029): Audit: reconnecting... (2)
2019/12/12 13:01:12 ossec-syscheckd[26029] syscheck_audit.c:1275 at audit_read_events(): INFO: (6030): Audit: connected.
2019/12/12 13:01:12 ossec-syscheckd[26029] syscheck_audit.c:1090 at audit_reload_rules(): DEBUG: (6275): Reloading Audit rules...
```

**Test 4**

Pytest's output:
```
[root@agent-master-1 test_fim]# python3 -m pytest test_audit/test_audit.py::test_move_folders_to_realtime
========================================== test session starts ===========================================
platform linux -- Python 3.7.4, pytest-5.3.1, py-1.8.0, pluggy-0.13.1
rootdir: /vagrant/wazuh-qa/test_wazuh, inifile: pytest.ini
collected 3 items                                                                                        

test_audit/test_audit.py .ss                                                                       [100%]

===================================== 1 passed, 2 skipped in 16.46s ======================================
[root@agent-master-1 test_fim]# 
```

Logs in `ossec.log`:
```
2019/12/12 15:07:40 ossec-syscheckd[32653] main.c:194 at main(): INFO: Started (pid: 32656).
2019/12/12 15:07:40 ossec-syscheckd[32653] main.c:202 at main(): INFO: (6003): Monitoring directory/file: '/testdir1', with options 'size | permissions | owner | group | mtime | inode | hash_md5 | hash_sha1 | hash_sha256 | whodata'.
2019/12/12 15:07:40 ossec-syscheckd[32653] main.c:202 at main(): INFO: (6003): Monitoring directory/file: '/testdir2', with options 'size | permissions | owner | group | mtime | inode | hash_md5 | hash_sha1 | hash_sha256 | whodata'.
2019/12/12 15:07:40 ossec-syscheckd[32653] main.c:202 at main(): INFO: (6003): Monitoring directory/file: '/testdir3', with options 'size | permissions | owner | group | mtime | inode | hash_md5 | hash_sha1 | hash_sha256 | whodata'.
2019/12/12 15:07:40 ossec-syscheckd[32653] syscom.c:111 at syscom_main(): DEBUG: (6284): Local requests thread ready.
2019/12/12 15:07:45 ossec-syscheckd[32653] syscheck_audit.c:428 at audit_init(): WARNING: (6224): Who-data engine cannot start because Auditd is not running.
2019/12/12 15:07:45 ossec-syscheckd[32653] main.c:254 at main(): WARNING: (6913): Who-data engine could not start. Switching who-data to real-time.
```

**Execute test_audit:**
```
[root@agent-master-1 test_fim]# python3 -m pytest test_audit
========================================== test session starts ===========================================
platform linux -- Python 3.7.4, pytest-5.3.1, py-1.8.0, pluggy-0.13.1
rootdir: /vagrant/wazuh-qa/test_wazuh, inifile: pytest.ini
collected 28 items                                                                                       

test_audit/test_audit.py ... ..sssssssss.ssssssss.                                                  [ 85%]
test_audit/test_auditd.py ..                                                                       [ 92%]
test_audit/test_remove_audit.py .                                                                  [ 96%]
test_audit/test_remove_rule_five_times.py .                                                        [100%]

=============================== 11 passed, 17 skipped in 155.62s (0:02:35) ===============================
[root@agent-master-1 test_fim]# 
```

### CODE REVIEW

**Comment by Lopuiz** [comment]
**Execute test on CentOS**
```
[root@agent-master-1 test_fim]# python3 -m pytest test_audit/
========================================== test session starts ===========================================
platform linux -- Python 3.7.4, pytest-5.3.1, py-1.8.0, pluggy-0.13.1
rootdir: /vagrant/wazuh-qa/test_wazuh, inifile: pytest.ini
collected 28 items                                                                                       

test_audit/test_audit.py .....sssssssss.ssssssss.                                                  [ 85%]
test_audit/test_audit_after_initial_scan.py ..                                                     [ 92%]
test_audit/test_remove_audit.py .                                                                  [ 96%]
test_audit/test_remove_rule_five_times.py .                                                        [100%]

=============================== 11 passed, 17 skipped in 163.43s (0:02:43) ===============================
[root@agent-master-1 test_fim]# 
```

**Execute test on Ubuntu**
```
root@agent:/wazuh/wazuh-qa/test_wazuh# python3 -m pytest test_fim/test_audit/
========================================== test session starts ===========================================
platform linux -- Python 3.7.3, pytest-5.3.1, py-1.8.0, pluggy-0.13.1
rootdir: /wazuh/wazuh-qa/test_wazuh, inifile: pytest.ini
collected 28 items                                                                                       

test_fim/test_audit/test_audit.py .....sssssssss.ssssssss.                                         [ 85%]
test_fim/test_audit/test_audit_after_initial_scan.py ..                                            [ 92%]
test_fim/test_audit/test_remove_audit.py .                                                         [ 96%]
test_fim/test_audit/test_remove_rule_five_times.py .                                               [100%]

=============================== 11 passed, 17 skipped in 159.81s (0:02:39) ===============================
root@agent:/wazuh/wazuh-qa/test_wazuh# 
```

File: test_wazuh/test_fim/test_audit/test_audit_after_initial_scan.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+    :param tags_to_apply Configuration tag to apply in the test
+    """
+
+    check_apply_test(tags_to_apply, get_configuration['tags'])
+
+    truncate_file(LOG_FILE_PATH)
+    wazuh_log_monitor = FileMonitor(LOG_FILE_PATH)
+
+    os.system("service auditd restart")
```
Comment (crd1985) [inline test_wazuh/test_fim/test_audit/test_audit_after_initial_scan.py]: Use `run` function from `subprocess` module instead.

File: test_wazuh/test_fim/test_audit/test_remove_audit.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    if re.match(linux_distro, "centos"):
+        package_management = "yum"
+        audit = "audit"
+        option = "--assumeyes"
+    elif re.match(linux_distro, "ubuntu") or re.match(linux_distro, "debian"):
+        package_management = "apt-get"
+        audit = "auditd"
+        option = "--yes"
+    else:
+        raise Exception("Can not uninstall/install audit")
```
Comment (crd1985) [inline test_wazuh/test_fim/test_audit/test_remove_audit.py]:
```suggestion
        raise ValueError(f"Linux distro ({linux_distro}) not supported for uninstall/install audit")
```

File: test_wazuh/test_fim/test_audit/test_remove_audit.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+    # Uninstall audit
+    process = subprocess.run([package_management, "remove", audit, option])
```
Comment (crd1985) [inline test_wazuh/test_fim/test_audit/test_remove_audit.py]:
```suggestion
    process = subprocess.run([package_management, "remove", audit, option], check=True)
```

File: test_wazuh/test_fim/test_audit/test_remove_audit.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    if process.returncode < 0:
+        raise Exception("Can not remove audit")
```
Comment (crd1985) [inline test_wazuh/test_fim/test_audit/test_remove_audit.py]: This check is performed in `subprocess.run` function if `check=True` is used.

File: test_wazuh/test_fim/test_audit/test_remove_audit.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+    yield
+
+    # Install audit and start the service
+    process = subprocess.run([package_management, "install", audit, option])
```
Comment (crd1985) [inline test_wazuh/test_fim/test_audit/test_remove_audit.py]:
```suggestion
    process = subprocess.run([package_management, "install", audit, option], check=True)
```

File: test_wazuh/test_fim/test_audit/test_remove_audit.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    if process.returncode < 0:
+        raise Exception("Can not install audit")
```
Comment (crd1985) [inline test_wazuh/test_fim/test_audit/test_remove_audit.py]: Same as above

File: test_wazuh/test_fim/test_audit/test_remove_audit.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+    process = subprocess.run(["service", "auditd", "start"])
```
Comment (crd1985) [inline test_wazuh/test_fim/test_audit/test_remove_audit.py]:
```suggestion
    process = subprocess.run(["service", "auditd", "start"], check=True)
```

File: test_wazuh/test_fim/test_audit/test_remove_audit.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    if process.returncode < 0:
+        raise Exception("Can not start audit")
```
Comment (crd1985) [inline test_wazuh/test_fim/test_audit/test_remove_audit.py]: Same as above

File: test_wazuh/test_fim/test_audit/test_audit_after_initial_scan.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+@pytest.fixture(scope='module', params=configurations)
+def get_configuration(request):
+    """Get configurations from the module."""
+    return request.param
+
+
+# Test
+
+@pytest.mark.parametrize('tags_to_apply, folder', [
+    ({'config1'}, '/testdir1')
```
Comment (crd1985) [inline test_wazuh/test_fim/test_audit/test_audit_after_initial_scan.py]: Maybe we could try also testdir2 and testdir3

File: test_wazuh/test_fim/test_audit/test_remove_rule_five_times.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+    :param tags_to_apply Configuration tag to apply in the test
+    :param folder The folder to remove and readd
+    :param audit_key The key which Wazuh put.
+    """
+
+    check_apply_test(tags_to_apply, get_configuration['tags'])
+
+    for i in range(0, 5):
+        os.system("auditctl -W " + folder + " -p wa -k " + audit_key)
```
Comment (crd1985) [inline test_wazuh/test_fim/test_audit/test_remove_rule_five_times.py]: Use `subprocess.run` as above

File: test_wazuh/test_fim/test_audit/test_audit_after_initial_scan.py (hunk 1 above)
Comment (Lopuiz) [reply test_wazuh/test_fim/test_audit/test_audit_after_initial_scan.py]: Done

File: test_wazuh/test_fim/test_audit/test_remove_audit.py (hunk 2 above)
Comment (Lopuiz) [reply test_wazuh/test_fim/test_audit/test_remove_audit.py]: Done

File: test_wazuh/test_fim/test_audit/test_remove_audit.py (hunk 3 above)
Comment (Lopuiz) [reply test_wazuh/test_fim/test_audit/test_remove_audit.py]: Done

File: test_wazuh/test_fim/test_audit/test_remove_audit.py (hunk 4 above)
Comment (Lopuiz) [reply test_wazuh/test_fim/test_audit/test_remove_audit.py]: Done

File: test_wazuh/test_fim/test_audit/test_remove_audit.py (hunk 5 above)
Comment (Lopuiz) [reply test_wazuh/test_fim/test_audit/test_remove_audit.py]: Done

File: test_wazuh/test_fim/test_audit/test_remove_audit.py (hunk 6 above)
Comment (Lopuiz) [reply test_wazuh/test_fim/test_audit/test_remove_audit.py]: Done

File: test_wazuh/test_fim/test_audit/test_remove_audit.py (hunk 7 above)
Comment (Lopuiz) [reply test_wazuh/test_fim/test_audit/test_remove_audit.py]: Done

File: test_wazuh/test_fim/test_audit/test_audit_after_initial_scan.py (hunk 9 above)
Comment (Lopuiz) [reply test_wazuh/test_fim/test_audit/test_audit_after_initial_scan.py]: Done

File: test_wazuh/test_fim/test_audit/test_remove_audit.py (hunk 8 above)
Comment (Lopuiz) [reply test_wazuh/test_fim/test_audit/test_remove_audit.py]: Done

File: test_wazuh/test_fim/test_audit/test_remove_rule_five_times.py (hunk 10 above)
Comment (Lopuiz) [reply test_wazuh/test_fim/test_audit/test_remove_rule_five_times.py]: Done

**Pull request merged**